### DISCUSSION TITLE
Evaluating the reward function for arbitrary states and actions

### DISCUSSION THREAD

**Discussion by jamesheald** [google/brax#326] [Q&A]
I'm implementing model predictive control and I want to be able to evaluate the reward associated with a rollout performed under a learned model. Is there a generic way (that works for any environment) to evaluate the reward associated with a state-action(-next state, if applicable) tuple?

**Comment by btaba** [comment]
If you're using the Env API, the reward would be stored in the State:
https://github.com/google/brax/blob/b373f5a45e62189a4a260131c17b10181ccda96a/brax/envs/env.py#L36

and if you want Transitions, you could use the actor_step function:
https://github.com/google/brax/blob/b373f5a45e62189a4a260131c17b10181ccda96a/brax/training/acting.py#L31-L51

**Reply by jamesheald** [reply]
Hi @btaba. I want to be able to access the reward function directly (and evaluate it for any observation [not full qp state which i'm not modelling] and action that I care to specify) not via the environment state, because I will be performing rollouts under a learned model, not under the environment step function. Does this make sense?

**Reply by btaba** [reply]
Hi @jamesheald, so it sounds like you are learning $f(s, a) = s'$, and want to evaluate a reward $r(s, s')$. You'll have to write a function $r$ to do that. We don't have any model-based examples in the codebase, but you can check out [acme agents](https://github.com/deepmind/acme/tree/master/acme/agents/jax) 

For example here is a planner implemented in jax:

https://github.com/deepmind/acme/blob/d1e69c92000079b118b868ce9303ee6d39c4a0b6/acme/agents/jax/mbop/mppi.py#L234

although I imagine you'll want something much simpler to start out with. Hope that helps

**Reply by jamesheald** [reply]
Thank you for your response. It would be useful, in a model-based RL context, to be able to evaluate the reward function without having to simulate a state transition via env.step (or write the reward function manually). I would like to test the model I'm developing on lots of different environments, and manually writing reward functions for all of them is a bit tedious (and potentially error prone). Maybe this feature could be added?

**Reply by btaba** [reply]
Hi @jamesheald, I'm not really understanding what the requested feature would be.  Are you trying to programmatically generate rewards for arbitrary environments? It sounds like you also don't want $r(s, s')$ but $r(s)$ (reward not dependent on the state transition)?

**Reply by jamesheald** [reply]
I would ideally like to be able to evaluate the (standard, pre-specified) reward function of any brax environment for any s, a and s' without having to (re)write the reward function myself. My understand is that currently the only way to get a reward from brax is to call env.step (with arguments s an r), which performs a state transition in the environment (returning s' and r).

**Reply by btaba** [reply]
Ok I think I understand! The most generic reward function is $r(s, a, s')$, and to get $s'$ one needs to calculate the env.step $f(s, a) = s'$. This is more or less why the reward function $r$ is rolled up in env.step, as found in other Env APIs as well such as [gymnasium](https://gymnasium.farama.org/api/env/) etc.
For model-based RL, I could see why you'd want a separate $r$. A short-term alternative is to override the `pipeline_step` function

https://github.com/google/brax/blob/b373f5a45e62189a4a260131c17b10181ccda96a/brax/envs/env.py#L116

to get the $s'$ you'd want the reward to be calculated for. Another alternative is to learn the reward along with the transition, as done in many model-based RL algorithms

**Reply by jamesheald** [reply]
Ok thanks for the response. In general, I think it would be nice to have a separate reward function that the user can call directly (and that env.step calls) for model-based scenarios where you only want to learn the transition dynamics but not the reward function.
